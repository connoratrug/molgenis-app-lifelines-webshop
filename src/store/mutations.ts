import ApplicationState from '@/types/applicationState'
import Variable from '@/types/Variable'

export default {
  updateGenderFilter (state: ApplicationState, selectedGenders: String[]) {
    state.facetFilter.gender = selectedGenders
  },
  updateSubcohortfilter (state: ApplicationState, selectedSubcohorts: String[]) {
    state.facetFilter.subcohort = selectedSubcohorts
  },
  updateSelectedAgeAt (state: ApplicationState, selectedAgeAt: any) {
    state.facetFilter.ageGroupAt1A = selectedAgeAt.ageGroupAt1A
    state.facetFilter.ageGroupAt2A = selectedAgeAt.ageGroupAt2A
    state.facetFilter.ageGroupAt3A = selectedAgeAt.ageGroupAt3A
  },
  updateYearOfBirthRangefilter (state: ApplicationState, yobRange: Number[]) {
    state.facetFilter.yearOfBirthRange = yobRange
  },
  removeYearOfBirthRangefilter (state: ApplicationState) {
    state.facetFilter.yearOfBirthRange = []
  },
  removeAgeAtFilter (state: ApplicationState) {
    state.facetFilter.ageGroupAt1A = []
    state.facetFilter.ageGroupAt2A = []
    state.facetFilter.ageGroupAt3A = []
  },
  updateTreeSelection (state: ApplicationState, selection: Number) {
    state.treeSelected = selection
  },
  updateSection (state: ApplicationState, sections: String[]) {
    state.sectionList = sections
  },
  updateTreeStructure (state: ApplicationState, sections: String[]) {
    state.treeStructure = sections
  },
  updateSubSection (state: ApplicationState, subSections: String[]) {
    state.subSectionList = subSections
  },
  updateVariables (state: ApplicationState, variables: Variable[]) {
    state.variables = variables
  }
}
