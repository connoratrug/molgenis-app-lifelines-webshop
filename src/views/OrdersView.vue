<template>

  <div id="orders-view" class="container pt-1">
    <ConfirmationModal
      v-if="$route && $route.name === 'orderDelete'"
      :backRoute="$router.resolve({name: 'orders'}).route"
      :confirmButton="$t('lifelines-webshop-modal-button-delete')"
      :confirmMethod="deleteOrderConfirmed.bind(this, $route.params.orderNumber)"
      :modalTitle="$t('lifelines-webshop-modal-delete-header', {order: $route.params.orderNumber})">
      {{$t('lifelines-webshop-modal-delete-body', {order: $route.params.orderNumber})}}
    </ConfirmationModal>

    <h1 id="orders-title">{{$t('lifelines-webshop-orders-title')}}</h1>

      <spinner-animation v-if="orders === null"/>
      <b-table v-else hover :items="orders" :fields="tableFields" per-page=5>

        <template v-slot:cell(actions)="data">
            <router-link
              tag="button"
              v-if="data.item.state === 'Draft' || hasManagerRole"
              class="btn btn-secondary btn-sm"
              :to="`/shop/${data.item.orderNumber}`">
                <font-awesome-icon icon="edit" aria-label="edit"/>
            </router-link>
            <button class="btn btn-secondary btn-sm copy-btn" type="button" @click="handleCopyOrder(data.item.orderNumber)">
              <font-awesome-icon icon="copy" aria-label="copy"/>
            </button>
            <router-link
              tag="button"
              v-if="data.item.state === 'Draft' || hasManagerRole"
              :to="{ name: 'orderDelete', params: {orderNumber: data.item.orderNumber}}"
              class="btn btn-danger btn-sm t-btn-order-delete">
              <font-awesome-icon icon="trash" aria-label="delete"/>
            </router-link>
        </template>

        <template v-slot:cell(submissionDate)="data">
           {{ data.item.submissionDate | dataString }}
        </template>

        <template v-slot:cell(applicationForm)="data">
          <a v-if="data.item.applicationForm" :href="data.item.applicationForm.url">
            {{ data.item.applicationForm.filename }} <font-awesome-icon icon="download" aria-label="download"/>
          </a>
        </template>

        <template v-slot:cell(status)="data">
          <b-dropdown id="dropdown-1" :text="data.item.state" class="m-md-2">
            <b-dropdown-item active>Draft</b-dropdown-item>
            <b-dropdown-item>Submitted</b-dropdown-item>
            <b-dropdown-item>Approved</b-dropdown-item>
          </b-dropdown>
        </template>
      </b-table>
    </div>
</template>

<script>
import Vue from 'vue'
import { library } from '@fortawesome/fontawesome-svg-core'
import { faEdit, faDownload, faTrash, faCopy } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import SpinnerAnimation from '../components/animations/SpinnerAnimation.vue'
import ConfirmationModal from '../components/ConfirmationModal.vue'
import { mapActions, mapState, mapMutations, mapGetters } from 'vuex'
import moment from 'moment'

library.add(faEdit, faDownload, faTrash, faCopy)

export default Vue.extend({
  components: { ConfirmationModal, FontAwesomeIcon, SpinnerAnimation },
  computed: {
    ...mapState(['orders']),
    ...mapGetters(['hasManagerRole']),
    tableFields: function () {
      let fields = [
        { key: 'actions', label: '', class: 'actions-column' }
      ]

      if (this.hasManagerRole) {
        fields.push({ key: 'email', label: this.$t('lifelines-webshop-orders-col-header-email'), sortable: true })
      }

      fields = fields.concat([
        { key: 'name', label: this.$t('lifelines-webshop-orders-col-header-title'), sortable: true },
        { key: 'submissionDate', label: this.$t('lifelines-webshop-orders-col-header-sub-date'), sortable: true },
        { key: 'projectNumber', label: this.$t('lifelines-webshop-orders-col-header-project'), sortable: true },
        { key: 'orderNumber', label: this.$t('lifelines-webshop-orders-col-header-order'), sortable: true },
        { key: 'applicationForm', label: this.$t('lifelines-webshop-orders-col-header-app-form') },
        { key: 'status', label: this.$t('lifelines-webshop-orders-col-header-state'), sortable: true }
      ])

      return fields
    }
  },
  data () {
    return {
      approvingOrder: '',
      perPage: 5
    }
  },
  methods: {
    deleteOrderConfirmed: function (orderNumber) {
      this.deleteOrder(orderNumber)
      this.$router.push({ name: 'orders' })
    },
    async handleApproveOrder (orderNumber) {
      this.approveState = orderNumber
      this.sendApproveTrigger(orderNumber).then(
        async () => {
          await this.loadOrders()
          this.approveState = ''
          this.setToast({ type: 'success', textType: 'light', title: 'Success', timeout: this.$global.toastTimeoutTime, message: `Order ${orderNumber} approved` })
        },
        () => {
          this.approveState = ''
          this.setToast({ type: 'danger', textType: 'light', message: `Order ${orderNumber} approval failed` })
        })
    },
    async handleCopyOrder (orderNumber) {
      await this.copyOrder(orderNumber)
      this.loadOrders()
    },
    ...mapActions(['loadOrders', 'deleteOrder', 'sendApproveTrigger', 'copyOrder']),
    ...mapMutations(['setToast'])
  },
  mounted () {
    this.loadOrders()
  },
  filters: {
    dataString: (dateValue) => dateValue ? moment(dateValue).format('LLLL') : ''
  }
})
</script>

<style lang="scss">
  @import "../scss/variables";

  .actions-column {
    width: 9rem;

    button {
      margin-right: $spacer;
    }
  }
</style>
