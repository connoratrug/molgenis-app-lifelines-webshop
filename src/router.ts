import Vue from 'vue'
import Router, { NavigationGuard } from 'vue-router'

import MainView from './views/MainView.vue'
import OrdersView from './views/OrdersView.vue'
import OrderView from './views/OrderView.vue'

import store from '@/store/store'

Vue.use(Router)
export const routes = [
  {
    path: '/',
    name: 'home',
    redirect: () => ({ name: store.getters.isSignedIn ? 'orders' : 'shop' })
  },
  {
    path: '/orders/:orderNumber?',
    name: 'orders',
    component: OrdersView,
    children: [
      {
        name: 'orderDelete',
        path: 'delete'
      }
    ]
  },
  {
    path: '/shop',
    name: 'shop',
    component: MainView
  },
  {
    path: '/shop/:orderNumber',
    name: 'load',
    component: MainView
  },
  {
    path: '/order',
    name: 'order',
    component: OrderView
  }
]

const packageJson = require('../package.json')

export const router = new Router({
  base: process.env.NODE_ENV === 'production' ? packageJson.name + '/dist/index.html' : process.env.BASE_URL,
  routes
})

export const anonymousNavigationGuard: NavigationGuard = (to, _from, next) => {
  if (to.name !== 'shop' && !store.getters.isSignedIn) {
    router.push({ name: 'shop' })
  } else {
    next()
  }
}
router.beforeResolve(anonymousNavigationGuard)
