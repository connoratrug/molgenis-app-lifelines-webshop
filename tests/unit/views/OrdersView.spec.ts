import { mount, shallowMount, createLocalVue } from '@vue/test-utils'
import Router from 'vue-router'
import { routes } from '@/router'
import OrdersView from '@/views/OrdersView.vue'
import moment from 'moment'
import Vuex from 'vuex'
import orders from '../fixtures/orders'
import Spinner from '@/components/animations/SpinnerAnimation.vue'
import mutations from '@/store/mutations'

describe('OrdersView.vue', () => {
  let localVue:any
  let store:any

  let actions = {
    deleteOrder: jest.fn(),
    loadOrders: jest.fn()
  }

  beforeEach(() => {
    localVue = createLocalVue()
    localVue.filter('moment', function (value: string, format: string) { return moment(value).utc().format(format) })
    localVue.filter('i18n', (value: string) => `#${value}#`)
    localVue.use(Vuex)

    let state:any = {
      orders: null
    }

    store = new Vuex.Store({
      state,
      actions,
      mutations
    })
  })

  it('renders orders view', () => {
    const wrapper = shallowMount(OrdersView, { store, localVue })
    expect(wrapper.find('#orders-view').exists()).toBeTruthy()
  })

  it('shows title', () => {
    const wrapper = shallowMount(OrdersView, { store, localVue })
    expect(wrapper.find('#orders-title').text()).toBe('lifelines-webshop-orders-title')
  })

  it('fetches orders when mounted', () => {
    shallowMount(OrdersView, { store, localVue })
    expect(actions.loadOrders).toHaveBeenCalled()
  })

  it('shows a spinner while loading orders', () => {
    const wrapper = shallowMount(OrdersView, { store, localVue })
    expect(wrapper.find(Spinner).isVisible()).toBeTruthy()
    store.commit('setOrders', orders)
    expect(wrapper.find(Spinner).exists()).toBeFalsy()
  })

  it('shows orders table when loaded', () => {
    const wrapper = shallowMount(OrdersView, { store, localVue })
    store.commit('setOrders', orders)
    expect(wrapper.find('table').isVisible()).toBeTruthy()
  })

  it('checks content of order tables', () => {
    const wrapper = shallowMount(OrdersView, { store, localVue })
    store.commit('setOrders', orders)
    expect(wrapper.find('table')).toMatchSnapshot()
  })

  it('shows a confirmation modal after pressing the delete button', () => {
    localVue.use(Router)

    const wrapper = mount(OrdersView, {
      localVue,
      store,
      router: new Router({ routes })
    })
    store.commit('setOrders', orders)

    expect(wrapper.find('.modal-dialog').exists()).toBe(false)
    wrapper.find('.t-btn-order-delete').trigger('click')
    expect(wrapper.find('.modal-dialog').exists()).toBe(true)
    wrapper.find('.t-btn-confirm').trigger('click')
    expect(actions.deleteOrder).toHaveBeenCalled()
  })
})
