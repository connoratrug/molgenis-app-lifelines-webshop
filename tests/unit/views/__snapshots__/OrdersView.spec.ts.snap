// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrdersView.vue checks content of order tables 1`] = `
<table aria-describedby="orders-title" class="table">
  <thead>
    <tr>
      <th scope="col"></th>
      <th scope="col"></th>
      <th scope="col">Title</th>
      <th scope="col">Submitted</th>
      <th scope="col">Project</th>
      <th scope="col">Order</th>
      <th scope="col">Application form</th>
      <th scope="col">State</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <router-link-stub to="/shop/edcba" tag="a" event="click" class="btn btn-primary btn-sm">
          <font-awesome-icon-stub icon="edit" aria-label="edit"></font-awesome-icon-stub>
        </router-link-stub>
      </td>
      <td>
        <router-link-stub to="[object Object]" tag="a" event="click" class="btn btn-danger btn-sm t-btn-order-delete">
          <font-awesome-icon-stub icon="trash" aria-label="delete"></font-awesome-icon-stub>
        </router-link-stub>
      </td>
      <td></td>
      <td></td>
      <td></td>
      <td>edcba</td>
      <td>
        <!---->
      </td>
      <td><span class="badge badge-pill badge-info">Draft</span></td>
    </tr>
    <tr>
      <td>
        <!---->
      </td>
      <td>
        <!---->
      </td>
      <td>My draft order</td>
      <td></td>
      <td></td>
      <td>abcde</td>
      <td><span><a href="https://lifelines.test.molgenis.org/files/aaaac3rcetmgfmudsodb3laaay">Motivation.pdf<font-awesome-icon-stub icon="download" aria-label="download"></font-awesome-icon-stub></a></span></td>
      <td><span class="badge badge-pill badge-primary">Submitted</span></td>
    </tr>
  </tbody>
</table>
`;
